File: bot/classifier.py
from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Protocol

from bot.groq import GroqEvidence
from bot.models import Classification, Confidence, Intent


class EvidenceClient(Protocol):
    def classify(self, text: str) -> GroqEvidence: ...

SIGNAL_SCORES: dict[str, tuple[Intent, int]] = {
    "explicit_need": ("lead", 4),
    "vendor_search": ("lead", 4),
    "pricing": ("lead", 2),
    "timeline": ("lead", 1),
    "contact_intent": ("lead", 2),
    "service_interest": ("lead", 1),
    "conversation": ("engagement", 2),
    "praise": ("engagement", 2),
    "criticism": ("engagement", 2),
    "promotion": ("spam", 4),
    "irrelevant": ("spam", 2),
}

LEAD_NEED_PHRASES = (
    "нужен сайт",
    "нужен лендинг",
    "нужен бот",
    "нужна разработка",
    "нужно приложение",
    "нужна автоматизация",
    "хочу сайт",
    "хочу лендинг",
    "хочу заказать",
    "хотим сайт",
    "нужно сделать сайт",
    "сделать сайт",
    "разработать сайт",
    "сайт керек",
    "қосымша керек",
)
VENDOR_SEARCH_PHRASES = (
    "ищу разработчика",
    "ищу подрядчика",
    "кто сделает сайт",
    "посоветуйте разработчика",
    "әзірлеуші іздеймін",
)
SERVICE_TERMS = (
    "сайт",
    "лендинг",
    "интернет-магазин",
    "приложение",
    "автоматизац",
    "crm",
    "бот",
    "дизайн",
    "разработ",
    "website",
    "app",
)
PRICING_PHRASES = ("сколько стоит", "какая цена", "стоимость", "цена разработки", "қанша тұрады")
TIMELINE_PHRASES = ("срочно", "на этой неделе", "за месяц", "срок", "дедлайн", "шұғыл")
CONTACT_PHRASES = ("напишите мне", "свяжитесь", "оставлю номер", "whatsapp", "телеграм", "telegram")
NEGATIONS = ("не нужен", "не нужна", "не нужно", "не ищу")
DIRECT_SERVICE_QUESTION = re.compile(
    r"(?:сколько\s+стоит|какая\s+цена|стоимость|какой\s+срок|как\s+заказать|что\s+входит|"
    r"что\s+нужно|чем\s+отличается|как\s+проходит|можно\s+(?:ли|подробнее)|"
    r"сможете|возьм[её]тесь|вы\s+(?:делаете|разрабатываете|собираете|настраиваете))[^?]{0,160}\?",
    re.IGNORECASE,
)
FIRST_PERSON_SERVICE_NEED = re.compile(
    r"\b(?:мне|нам|мы|я|хочу|хотим|планирую|планируем|у\s+нас)\b[^.!?]{0,100}"
    r"(?:сайт|лендинг|интернет-магазин|приложени\w*|автоматизац\w*|crm|бот\w*|дизайн|разработк\w*)",
    re.IGNORECASE,
)
SPAM_PHRASES = (
    "заработок без вложений",
    "крипто сигнал",
    "гарантированный доход",
    "подпишись на канал",
    "накрутка подписчиков",
    "casino",
    "казино",
)
RISK_PATTERNS: dict[str, tuple[str, ...]] = {
    "aggression": ("идиот", "тупые", "ненавижу", "заткнись", "уроды"),
    "complaint": ("жалоба", "обманули", "мошенники", "верните деньги", "ужасный сервис"),
    "legal": ("подам в суд", "юрист", "претензия", "нарушение закона", "судеб"),
    "reputation": ("опубликую отзыв", "разнесу в соцсетях", "репутац"),
}
CRITICISM_PHRASES = (
    "не согласен",
    "не согласна",
    "ерунда",
    "бред",
    "чушь",
    "глупость",
    "неправда",
    "сомнительно",
    "автор не понимает",
    "вы не понимаете",
)
ENGAGEMENT_REPLY_BLOCKED = re.compile(
    r"(?:whatsapp|telegram|wa\.me|напиш\w*\s+(?:нам|мне|в)|остав\w*\s+(?:номер|заявк)|"
    r"закаж\w*|стоимост|цена|спасибо\s+за\s+(?:обратную\s+связь|ваше\s+мнение)|"
    r"благодарим|рады,\s+что|обращайтесь)",
    re.IGNORECASE,
)


def _contains_any(text: str, phrases: Iterable[str]) -> bool:
    return any(phrase in text for phrase in phrases)


def _local_signals(text: str) -> tuple[set[str], set[str]]:
    normalized = re.sub(r"\s+", " ", text.casefold()).strip()
    signals: set[str] = set()
    risks = {flag for flag, patterns in RISK_PATTERNS.items() if _contains_any(normalized, patterns)}
    negated = _contains_any(normalized, NEGATIONS)

    if not negated and _contains_any(normalized, LEAD_NEED_PHRASES):
        signals.add("explicit_need")
    if not negated and _contains_any(normalized, VENDOR_SEARCH_PHRASES):
        signals.add("vendor_search")
    if _contains_any(normalized, SERVICE_TERMS):
        signals.add("service_interest")
    if _contains_any(normalized, PRICING_PHRASES):
        signals.add("pricing")
    if _contains_any(normalized, TIMELINE_PHRASES):
        signals.add("timeline")
    if _contains_any(normalized, CONTACT_PHRASES):
        signals.add("contact_intent")
    if _contains_any(normalized, SPAM_PHRASES):
        signals.add("promotion")
    if "?" in text or normalized.startswith(("как ", "что ", "почему ", "спасибо", "класс", "интересно")):
        signals.add("conversation")
    if _contains_any(normalized, ("класс", "круто", "отлично", "полезно", "спасибо", "супер")):
        signals.add("praise")
    if _contains_any(normalized, CRITICISM_PHRASES):
        signals.add("criticism")
    return signals, risks


def is_direct_commercial_message(text: str) -> bool:
    signals, _ = _local_signals(text)
    if {"explicit_need", "vendor_search"} & signals:
        return True
    if FIRST_PERSON_SERVICE_NEED.search(text):
        return True
    if "service_interest" in signals and DIRECT_SERVICE_QUESTION.search(text):
        return True
    return "?" in text and bool({"pricing", "timeline", "contact_intent"} & signals)


def _scores(signals: Iterable[str]) -> dict[Intent, int]:
    result: dict[Intent, int] = {"lead": 0, "engagement": 0, "spam": 0}
    for signal in signals:
        scoring = SIGNAL_SCORES.get(signal)
        if scoring:
            intent, points = scoring
            result[intent] += points
    return result


def _confidence(scores: dict[Intent, int], winner: Intent) -> Confidence:
    ordered = sorted(scores.values(), reverse=True)
    top = scores[winner]
    margin = top - ordered[1]
    if top >= 5 and margin >= 2:
        return "high"
    if top >= 3 and margin >= 1:
        return "medium"
    return "low"


class Classifier:
    def __init__(self, groq: EvidenceClient, whatsapp_contact_link: str = "") -> None:
        self.groq = groq
        self.whatsapp_contact_link = whatsapp_contact_link.strip()

    def classify(self, text: str) -> Classification:
        local_signals, local_risks = _local_signals(text)
        local_scores = _scores(local_signals)

        if local_risks:
            risk_intent: Intent = max(local_scores, key=local_scores.get) if any(local_scores.values()) else "engagement"
            return Classification(
                intent=risk_intent,
                signals=tuple(sorted(local_signals)),
                risk_flags=tuple(sorted(local_risks)),
                confidence_level="low",
                bot_reply_text=None,
            )
        if local_scores["spam"] >= 4:
            return self._result("spam", local_signals, (), "high", None)
        if local_scores["lead"] >= 5:
            return self._result("lead", local_signals, (), "high", self._lead_reply())
        if local_scores["lead"] >= 3 and local_scores["lead"] > local_scores["engagement"]:
            return self._result("lead", local_signals, (), "medium", self._lead_reply())
        if local_scores["engagement"] >= 4 and local_scores["lead"] == 0:
            reply = None if "criticism" in local_signals else self._engagement_reply(text)
            return self._result(
                "engagement",
                local_signals,
                (),
                "high",
                reply,
            )

        evidence = self.groq.classify(text)
        direct_commercial_message = is_direct_commercial_message(text)
        evidence_signals = set(evidence.signals)
        if not direct_commercial_message:
            evidence_signals -= {"explicit_need", "vendor_search", "contact_intent"}
        signals = local_signals | evidence_signals
        risks = local_risks | set(evidence.risk_flags)
        if not direct_commercial_message and evidence.intent != "spam":
            signals.add("conversation")
            reply = None
            if "criticism" not in signals and evidence.intent == "engagement":
                reply = self._safe_engagement_reply(evidence.proposed_reply)
            return self._result("engagement", signals, risks, "high", reply)
        scores = _scores(signals)
        scores[evidence.intent] += 1
        winner = max(scores, key=scores.get)
        confidence = _confidence(scores, winner)
        reply = None
        if not risks:
            if winner == "lead":
                reply = self._lead_reply()

        return self._result(winner, signals, risks, confidence, reply)

    def _lead_reply(self) -> str:
        if self.whatsapp_contact_link:
            return (
                "Похоже, здесь можем помочь. Напишите пару деталей в WhatsApp — "
                f"посмотрим задачу без навязчивых продаж: {self.whatsapp_contact_link}"
            )
        return "Похоже, здесь можем помочь. Напишите пару деталей — спокойно посмотрим задачу."

    def _engagement_reply(self, text: str) -> str | None:
        try:
            evidence = self.groq.classify(text)
        except Exception:
            return None
        if (
            evidence.intent != "engagement"
            or evidence.risk_flags
            or "criticism" in evidence.signals
        ):
            return None
        return self._safe_engagement_reply(evidence.proposed_reply)

    @staticmethod
    def _safe_engagement_reply(reply: str | None) -> str | None:
        if not reply:
            return None
        normalized = reply.strip()
        if len(normalized) > 180 or ENGAGEMENT_REPLY_BLOCKED.search(normalized):
            return None
        return normalized

    @staticmethod
    def _result(
        intent: Intent,
        signals: Iterable[str],
        risks: Iterable[str],
        confidence: Confidence,
        reply: str | None,
    ) -> Classification:
        return Classification(
            intent=intent,
            signals=tuple(sorted(set(signals))),
            risk_flags=tuple(sorted(set(risks))),
            confidence_level=confidence,
            bot_reply_text=reply,
        )

File: supabase/functions/_shared/classifier.ts
import type { GroqClient } from "./groq.ts";
import type { Classification, ConfidenceLevel, Intent } from "./types.ts";

const SIGNAL_SCORES: Record<string, [Intent, number]> = {
  explicit_need: ["lead", 4],
  vendor_search: ["lead", 4],
  pricing: ["lead", 2],
  timeline: ["lead", 1],
  contact_intent: ["lead", 2],
  service_interest: ["lead", 1],
  conversation: ["engagement", 2],
  praise: ["engagement", 2],
  criticism: ["engagement", 2],
  promotion: ["spam", 4],
  irrelevant: ["spam", 2],
};

const LEAD_NEED_PHRASES = [
  "нужен сайт",
  "нужен лендинг",
  "нужен бот",
  "нужна разработка",
  "нужно приложение",
  "нужна автоматизация",
  "хочу сайт",
  "хочу лендинг",
  "хочу заказать",
  "хотим сайт",
  "нужно сделать сайт",
  "сделать сайт",
  "разработать сайт",
  "сайт керек",
  "қосымша керек",
];
const VENDOR_SEARCH_PHRASES = [
  "ищу разработчика",
  "ищу подрядчика",
  "кто сделает сайт",
  "посоветуйте разработчика",
  "әзірлеуші іздеймін",
];
const SERVICE_TERMS = [
  "сайт",
  "лендинг",
  "интернет-магазин",
  "приложение",
  "автоматизац",
  "crm",
  "бот",
  "дизайн",
  "разработ",
  "website",
  "app",
];
const PRICING_PHRASES = [
  "сколько стоит",
  "какая цена",
  "стоимость",
  "цена разработки",
  "қанша тұрады",
];
const TIMELINE_PHRASES = ["срочно", "на этой неделе", "за месяц", "срок", "дедлайн", "шұғыл"];
const CONTACT_PHRASES = [
  "напишите мне",
  "свяжитесь",
  "оставлю номер",
  "whatsapp",
  "телеграм",
  "telegram",
];
const NEGATIONS = ["не нужен", "не нужна", "не нужно", "не ищу"];
const DIRECT_SERVICE_QUESTION =
  /(?:сколько\s+стоит|какая\s+цена|стоимость|какой\s+срок|как\s+заказать|что\s+входит|что\s+нужно|чем\s+отличается|как\s+проходит|можно\s+(?:ли|подробнее)|сможете|возьм[её]тесь|вы\s+(?:делаете|разрабатываете|собираете|настраиваете))[^?]{0,160}\?/iu;
const FIRST_PERSON_SERVICE_NEED =
  /(?:^|[^\p{L}])(?:мне|нам|мы|я|хочу|хотим|планирую|планируем|у\s+нас)(?:$|[^\p{L}])[^.!?]{0,100}(?:сайт|лендинг|интернет-магазин|приложени\p{L}*|автоматизац\p{L}*|crm|бот\p{L}*|дизайн|разработк\p{L}*)/iu;
const SPAM_PHRASES = [
  "заработок без вложений",
  "крипто сигнал",
  "гарантированный доход",
  "подпишись на канал",
  "накрутка подписчиков",
  "casino",
  "казино",
];
const RISK_PATTERNS: Record<string, string[]> = {
  aggression: ["идиот", "тупые", "ненавижу", "заткнись", "уроды"],
  complaint: ["жалоба", "обманули", "мошенники", "верните деньги", "ужасный сервис"],
  legal: ["подам в суд", "юрист", "претензия", "нарушение закона", "судеб"],
  reputation: ["опубликую отзыв", "разнесу в соцсетях", "репутац"],
};
const CRITICISM_PHRASES = [
  "не согласен",
  "не согласна",
  "ерунда",
  "бред",
  "чушь",
  "глупость",
  "неправда",
  "сомнительно",
  "автор не понимает",
  "вы не понимаете",
];

function containsAny(text: string, phrases: readonly string[]): boolean {
  return phrases.some((phrase) => text.includes(phrase));
}

export function localEvidence(text: string): { signals: Set<string>; risks: Set<string> } {
  const normalized = text.toLocaleLowerCase().replace(/\s+/g, " ").trim();
  const signals = new Set<string>();
  const risks = new Set<string>();
  for (const [flag, patterns] of Object.entries(RISK_PATTERNS)) {
    if (containsAny(normalized, patterns)) risks.add(flag);
  }
  const negated = containsAny(normalized, NEGATIONS);

  if (!negated && containsAny(normalized, LEAD_NEED_PHRASES)) signals.add("explicit_need");
  if (!negated && containsAny(normalized, VENDOR_SEARCH_PHRASES)) signals.add("vendor_search");
  if (containsAny(normalized, SERVICE_TERMS)) signals.add("service_interest");
  if (containsAny(normalized, PRICING_PHRASES)) signals.add("pricing");
  if (containsAny(normalized, TIMELINE_PHRASES)) signals.add("timeline");
  if (containsAny(normalized, CONTACT_PHRASES)) signals.add("contact_intent");
  if (containsAny(normalized, SPAM_PHRASES)) signals.add("promotion");
  if (
    text.includes("?") ||
    ["как ", "что ", "почему ", "спасибо", "класс", "интересно"].some((prefix) =>
      normalized.startsWith(prefix)
    )
  ) signals.add("conversation");
  if (containsAny(normalized, ["класс", "круто", "отлично", "полезно", "спасибо", "супер"])) {
    signals.add("praise");
  }
  if (containsAny(normalized, CRITICISM_PHRASES)) signals.add("criticism");
  return { signals, risks };
}

export function isDirectCommercialMessage(text: string): boolean {
  const { signals } = localEvidence(text);
  if (signals.has("explicit_need") || signals.has("vendor_search")) return true;
  if (FIRST_PERSON_SERVICE_NEED.test(text)) return true;
  if (signals.has("service_interest") && DIRECT_SERVICE_QUESTION.test(text)) return true;

  const asksCommercialDetail = text.includes("?") &&
    (signals.has("pricing") || signals.has("timeline") || signals.has("contact_intent"));
  return asksCommercialDetail;
}

function scores(signals: Iterable<string>): Record<Intent, number> {
  const result: Record<Intent, number> = { lead: 0, engagement: 0, spam: 0 };
  for (const signal of signals) {
    const scoring = SIGNAL_SCORES[signal];
    if (scoring) result[scoring[0]] += scoring[1];
  }
  return result;
}

function winningIntent(values: Record<Intent, number>): Intent {
  return (Object.entries(values) as Array<[Intent, number]>).reduce(
    (best, current) => current[1] > best[1] ? current : best,
  )[0];
}

function confidence(values: Record<Intent, number>, winner: Intent): ConfidenceLevel {
  const ordered = Object.values(values).sort((left, right) => right - left);
  const top = values[winner];
  const margin = top - ordered[1];
  if (top >= 5 && margin >= 2) return "high";
  if (top >= 3 && margin >= 1) return "medium";
  return "low";
}

function sorted(values: Iterable<string>): string[] {
  return [...new Set(values)].sort();
}

export class Classifier {
  constructor(
    private readonly groq: Pick<GroqClient, "classify">,
    private readonly whatsappContactLink = "",
    private readonly businessContext = "",
  ) {}

  async classify(text: string): Promise<Classification> {
    const { signals: localSignals, risks: localRisks } = localEvidence(text);
    const localScores = scores(localSignals);

    if (localRisks.size > 0) {
      const riskIntent = Object.values(localScores).some(Boolean)
        ? winningIntent(localScores)
        : "engagement";
      return this.result(riskIntent, localSignals, localRisks, "low", null);
    }
    if (localScores.spam >= 4) return this.result("spam", localSignals, [], "high", null);
    if (localScores.lead >= 5) {
      return this.result("lead", localSignals, [], "high", await this.leadReply(text));
    }
    if (localScores.lead >= 3 && localScores.lead > localScores.engagement) {
      return this.result("lead", localSignals, [], "medium", await this.leadReply(text));
    }
    if (localScores.engagement >= 4 && localScores.lead === 0) {
      const reply = localSignals.has("criticism") ? null : await this.engagementReply(text);
      return this.result("engagement", localSignals, [], "high", reply);
    }

    const evidence = await this.groq.classify(text, this.businessContext);
    const directCommercialMessage = isDirectCommercialMessage(text);
    const evidenceSignals = directCommercialMessage
      ? evidence.signals
      : evidence.signals.filter((signal) =>
        signal !== "explicit_need" && signal !== "vendor_search" && signal !== "contact_intent"
      );
    const combinedSignals = new Set([...localSignals, ...evidenceSignals]);
    const combinedRisks = new Set([...localRisks, ...evidence.riskFlags]);
    if (!directCommercialMessage && evidence.intent !== "spam") {
      combinedSignals.add("conversation");
      const reply = combinedSignals.has("criticism") || evidence.intent !== "engagement"
        ? null
        : evidence.proposedReply;
      return this.result(
        "engagement",
        combinedSignals,
        combinedRisks,
        "high",
        reply,
      );
    }
    const combinedScores = scores(combinedSignals);
    combinedScores[evidence.intent] += 1;
    const winner = winningIntent(combinedScores);
    let reply: string | null = null;
    if (combinedRisks.size === 0 && winner === "lead") {
      reply = this.withContact(evidence.proposedReply ?? this.defaultLeadReply());
    }

    return this.result(
      winner,
      combinedSignals,
      combinedRisks,
      confidence(combinedScores, winner),
      reply,
    );
  }

  private async leadReply(text: string): Promise<string> {
    if (this.businessContext.trim()) {
      try {
        const evidence = await this.groq.classify(text, this.businessContext);
        if (
          evidence.intent === "lead" &&
          evidence.riskFlags.length === 0 &&
          evidence.proposedReply
        ) {
          return this.withContact(evidence.proposedReply);
        }
      } catch {
        console.warn(JSON.stringify({ event: "personalized_reply_fallback" }));
      }
    }
    return this.withContact(this.defaultLeadReply());
  }

  private async engagementReply(text: string): Promise<string | null> {
    try {
      const evidence = await this.groq.classify(text, this.businessContext);
      if (
        evidence.intent === "engagement" &&
        evidence.riskFlags.length === 0 &&
        !evidence.signals.includes("criticism") &&
        evidence.proposedReply
      ) {
        return evidence.proposedReply;
      }
    } catch {
      console.warn(JSON.stringify({ event: "engagement_reply_fallback" }));
    }
    return null;
  }

  private defaultLeadReply(): string {
    return "Похоже, здесь можем помочь. Напишите пару деталей — спокойно посмотрим задачу.";
  }

  private withContact(reply: string): string {
    const link = this.whatsappContactLink.trim();
    const normalized = reply.trim();
    if (!link || normalized.includes(link)) return Array.from(normalized).slice(0, 450).join("");

    const suffix = ` Если удобно, детали можно отправить в WhatsApp: ${link}`;
    const available = Math.max(0, 450 - Array.from(suffix).length);
    const prefix = Array.from(normalized).slice(0, available).join("").trimEnd();
    return `${prefix}${suffix}`;
  }

  private result(
    intent: Intent,
    signals: Iterable<string>,
    risks: Iterable<string>,
    confidenceLevel: ConfidenceLevel,
    botReplyText: string | null,
  ): Classification {
    return {
      intent,
      signals: sorted(signals),
      riskFlags: sorted(risks),
      confidenceLevel,
      botReplyText,
    };
  }
}
